python/ray/data/dataset_pipeline.py

```diff
@@ -1123,8 +1123,15 @@ def from_iterable(
         return DatasetPipeline(iterable, False, length=length)
 
     def __repr__(self) -> str:
-        return "DatasetPipeline(num_windows={}, num_stages={})".format(
-            self._length, 1 + len(self._stages)
+        if self._executed[0]:
+            return (
+                f"DatasetPipeline(num_windows={self._length}, "
+                f"num_stages_unoptimized={len(self._stages)}, "
+                f"num_stages_optimized={len(self._optimized_stages)})"
```

Comment:
What is an optimized/unoptimized stage?

Reply:
AFAIK optimized stage is stage by fusing multiple original stages together. btw why we need to expose the information here? curious what's the use case?

Reply:
Yeah my confusion here is about how to interpret this as a generic user. Some thoughts:

1. It's not clear what is counted as an unoptimized stage (i.e. is it the original number of stages or the stages that are yet to be optimized).
2. This depends on whether or not stages have been executed to trigger optimization.
3. Should this information be available in `Dataset`s as well?

Reply:
How about just calling this `num_stages` and `num_stages_fused`? I don't think we use the term optimize but do say fusion.

Reply:
Hmm seems the strange part is that we don't show the stages after optimization in most cases (user prints pipeline prior to execution).

Is it possible to run optimization before generating the repr? That hopefully shouldn't be a big overhead.

Reply:
I'm just studying this magic method and `__str__' in Python, what I see the convention is:
- `__repr__` is for developers (who develop the lib): it should capture what's deciding this class ("unambiguous"). In this case, it should be base_datasets and stages. Similar for Dataset, it's input_blocks and plan.
- `__str__` is for users (who use the lib): it should make it more readable and user friendly, such as schema, num of windows (I think stages is optional to mention). For Dataset, it's current `__repr__` which print schema and num of rows etc (which are user oriented).

Reply:
Narrow down the change to just fix the existing num_stages. PTAL.